### DISCUSSION TITLE
[RFC] Allow to gracefully shutdown persistent tasks

### DISCUSSION THREAD

**Discussion by gajus** [vercel/turborepo#7479] [Ideas]
### Which project is this feature idea for?

Turborepo

### Describe the feature you'd like to request

At the moment, as soon `turbo` receives SIGINT it kills the child process.

### Describe the solution you'd like

I would like to be able to gracefully handle the shutdown.

### Describe alternatives you've considered

Not using turbo

**Comment by gajus** [comment]
This issue is equivalent to https://github.com/pnpm/pnpm/issues/2653#issuecomment-1476686711

**Comment by gajus** [comment]
Appears like this is a regression

https://github.com/vercel/turbo/issues/3711

**Comment by williamlmao** [comment]
Lately I have been having issues where hitting command + c will appear to stop the process in the terminal, but the app will still be running on the port. Is this the same issue you are referring to @gajus ?

**Comment by arlyon** [comment]
@williamlmao we had a regression in 1.8.4 released yesterday which was fixed in 1.8.5 released earlier today, could you try that?

**Comment by gajus** [comment]
I was using 1.8.3 when posting this, so probably a distinct issue from what @arlyon is referring to.

Mine is about waiting for the script to actually complete before force exiting turbo.

In ideal world, the way this would work is that the first time `SIGINT` is received, `turbo` passes the signal to the child processes and waits for them to exit, and only if another `SIGINT` is received that `turbo` force exits.

**Comment by chris-olszewski** [comment]
@gajus How are you invoking turbo? I've been digging into this and have been getting vastly different results depending on if you invoke turbo globally or via a package manager.

**Comment by gajus** [comment]
Yeah, pnpm has the same flaw

https://github.com/pnpm/pnpm/issues/2653#issuecomment-1476686711

I was reporting what the behaviour is if directly invoking turbo though, i.e. `./node_modules/.bin/turbo`

**Comment by BarakChamo** [comment]
Chiming in here with the same issue in a simple launch script that we need to run with turbo in persistent mode.
The reason for having a launch script is to run a `docker-compose` setup, but terminate it when turbo exits.

This is the result of running the script directly in the terminal from the monorepo with `bun run up` (at exit):
```bash
barakchamo@barak store % Stopping local Postgres
[+] Stopping 2/2
 ✔ Container store-pg_proxy-1  Stopped                                                                                                                                                                                                                                                                                             
 ✔ Container store-postgres-1  Stopped                                                                                                                                                                                                                                                                                             
canceled
[+] Running 3/0
 ✔ Container store-pg_proxy-1  Removed                                                                                                                                                                                                                                                                                             
 ✔ Container store-postgres-1  Removed                                                                                                                                                                                                                                                                                             
 ✔ Network store_default       Removed    
```

And this is the result of running the script from the monorepo root via turbo `turbo run up` (at exiting):
```
@this/store:up: Stopping local Postgres

barakchamo@barak bounce %  ERROR  run failed: command  exited (1)
```

There's no issue running the launch script with a package manager, 
it's the execution from turbo that is causing the early termination before cleanup is completed.

**Comment by melroy89** [comment]
I'm running Fastify with [graceful shutdown plugin](https://www.npmjs.com/package/fastify-graceful-shutdown). Works great!

However, when using Turbo and using ctrl+c to close the dev environment for example. Turbo seems to **hard kill** the Turbo tasks _instead_ of gracefully shutdown instead.

```
 ERROR  run failed: command  exited (1)
 ELIFECYCLE  Command failed with exit code 1.
```

This is still an issue in 2025.

**Comment by lukasvan3l** [comment]
In 2026 as well :(

I'm running into this issue because one of my apps is a Firebase Functions with an emulator. If I open the emulator in a separate terminal and kill it with ctrl+c, the emulator saves its data to disk. If the emulator is one of the turbo processes, it gets killed too fast for the emulator to save data to disk.

If you need a reproducable repo i can create one with a turborepo app and a firebase project in it.

```
CLI:
   Version: 2.7.6
   Path to executable: /Users/xxx/Projects/xxx/node_modules/turbo-darwin-arm64/bin/turbo
   Daemon status: Running
   Package manager: npm

Platform:
   Architecture: aarch64
   Operating system: macos
   WSL: false
   Available memory (MB): 11107
   Available CPU cores: 10

Environment:
   CI: None
   Terminal (TERM): xterm-256color
   Terminal program (TERM_PROGRAM): vscode
   Terminal program version (TERM_PROGRAM_VERSION): 2.4.21
   Shell (SHELL): /bin/zsh
   stdin: false
```
